src/Sentry/Protocol/Metrics/Metric.cs

```diff
@@ -0,0 +1,126 @@
+using Sentry.Extensibility;
+using Sentry.Internal.Extensions;
+using ISentrySerializable = Sentry.Protocol.Envelopes.ISerializable;
+
+namespace Sentry.Protocol.Metrics;
+
+internal abstract class Metric : IJsonSerializable, ISentrySerializable
+{
+    protected Metric() : this(string.Empty)
+    {
+    }
+
+    protected Metric(string key, MeasurementUnit? unit = null, IDictionary<string, string>? tags = null, DateTime? timestamp = null)
+    {
+        Key = key;
+        Unit = unit;
+        Tags = tags ?? new Dictionary<string, string>();
+        Timestamp = timestamp ?? DateTime.UtcNow;
+    }
+
+    public SentryId EventId { get; private set; } = SentryId.Create();
+
+    public string Key { get; private set; }
+
+    public DateTime Timestamp { get; private set; }
+
+    public MeasurementUnit? Unit { get; private set; }
+
+    public IDictionary<string, string> Tags { get; private set; }
+
+    public abstract void Add(double value);
+
+    protected abstract void WriteValues(Utf8JsonWriter writer, IDiagnosticLogger? logger);
+
+    public void WriteTo(Utf8JsonWriter writer, IDiagnosticLogger? logger)
+    {
+        writer.WriteStartObject();
+        writer.WriteString("type", GetType().Name);
+        writer.WriteSerializable("event_id", EventId, logger);
+        writer.WriteString("name", Key);
+        writer.WriteString("timestamp", Timestamp);
+        if (Unit.HasValue)
+        {
+            writer.WriteStringIfNotWhiteSpace("unit", Unit.ToString());
+        }
+        writer.WriteStringDictionaryIfNotEmpty("tags", Tags!);
+        WriteValues(writer, logger);
+        writer.WriteEndObject();
+    }
+
+    protected abstract IEnumerable<IConvertible> SerializedStatsdValues();
+
+    internal static string SanitizeKey(string input) => Regex.Replace(input, @"[^a-zA-Z0-9_/.-]+", "_");
+    internal static string SanitizeValue(string input) => Regex.Replace(input, @"[^\w\d_:/@\.\{\}\[\]$-]+", "_");
+
+    public async Task SerializeAsync(Stream stream, IDiagnosticLogger? logger, CancellationToken cancellationToken = default)
+    {
+        /*
+         * We're serializing using the statsd format here: https://github.com/b/statsd_spec
+         */
+        var metricName = SanitizeKey(Key);
+        await Write(metricName).ConfigureAwait(false);
```

Comment:
afaik `.ConfigureAwait(false);` after the first one of these it's already didn't capture ExecutionContext so we could remove all the ones below. We do have some linter that yells at us but in thsi case since there are so many might be easier to disable it for the rest of the method instead of complying to it by adding this everywhere.